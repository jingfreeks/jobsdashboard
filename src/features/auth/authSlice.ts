import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

interface AuthState {
  user: null | { email: string; name?: string };
  loading: boolean;
  error: string | null;
}

const initialState: AuthState = {
  user: null,
  loading: false,
  error: null,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    loginStart(state) {
      state.loading = true;
      state.error = null;
    },
    loginSuccess(state, action: PayloadAction<{ email: string; name?: string }>) {
      state.user = action.payload;
      state.loading = false;
      state.error = null;
    },
    loginFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.error = action.payload;
    },
    registerStart(state) {
      state.loading = true;
      state.error = null;
    },
    registerSuccess(state, action: PayloadAction<{ email: string; name?: string }>) {
      state.user = action.payload;
      state.loading = false;
      state.error = null;
    },
    registerFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.error = action.payload;
    },
    logout(state) {
      state.user = null;
      state.loading = false;
      state.error = null;
    },
  },
});

export const {
  loginStart,
  loginSuccess,
  loginFailure,
  registerStart,
  registerSuccess,
  registerFailure,
  logout,
} = authSlice.actions;

export default authSlice.reducer; 