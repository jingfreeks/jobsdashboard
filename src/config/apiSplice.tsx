/* eslint-disable @typescript-eslint/no-explicit-any */
import {setCredentials, setLogout} from '@/features/auth';
import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
const baseUrl = import.meta.env.VITE_API_BASE_URL;
const baseQuery = fetchBaseQuery({
  baseUrl,
  credentials: 'include',
   
  prepareHeaders: (headers, {getState}: any) => {
    const token = getState().auth.token;
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    return headers;
  },
});

const baseQueryWithAuth = async (
  args: string,
  api: any,
  extraOptions: string,
) => {
  let result: any = await baseQuery(args, api, extraOptions);
  console.log(result);
  if (result?.error?.originalStatus === 403) {
    const refereshResult = await baseQuery('/refresh', api, extraOptions);

    if (refereshResult?.data) {
      const currentAuth = api.getState().auth;
      //store the new token
      api.dispatch(setCredentials({
        user: currentAuth.user,
        accessToken: (refereshResult.data as any).accessToken,
        userId: currentAuth.userId,
        roles: currentAuth.roles
      }));
      //retry the original query with new access token
      result = await baseQuery(args, api, extraOptions);
    } else {
      api.dispatch(setLogout());
    }
  }
  return result;
};

export const apiSlice = createApi<any, any>({
  baseQuery: baseQueryWithAuth,
  reducerPath: 'api',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  endpoints: _ => ({}),
  tagTypes: ['City', 'Jobs', 'MyJobs', 'JobDetails', 'Profile'] as any,
});