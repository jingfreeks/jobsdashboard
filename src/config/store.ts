import { configureStore } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from '@/features/auth';
import {apiSlice} from './apiSplice';

// Configure persistence for auth reducer
const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['user', 'token', 'userId', 'roles'], // Only persist these fields
  timeout: 10000, // 10 second timeout for storage operations
  serialize: true, // Enable serialization
  deserialize: true, // Enable deserialization
};

const persistedAuthReducer = persistReducer(authPersistConfig, authReducer);

export const store = configureStore({
  reducer: {
    auth: persistedAuthReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: ['auth'], // Ignore auth state for serialization checks
      },
      immutableCheck: {
        ignoredPaths: ['auth'], // Ignore auth state for immutability checks
      },
    }).concat(apiSlice.middleware),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch; 