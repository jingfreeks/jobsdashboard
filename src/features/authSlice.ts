import {apiSlice} from '@/config/apiSplice';

export const authApiSlice = apiSlice.injectEndpoints({
  endpoints: builder => ({
    login: builder.mutation({
      query: credentials => ({
        url: '/auth',
        method: 'POST',
        body: {...credentials},
      }),
    }),
    logout: builder.mutation({
      query: () => ({
        url: '/logout',
        method: 'POST',
        // body: {...credentials},
      }),
    }),
    signup: builder.mutation({
      query: credentials => ({
        url: '/signup',
        method: 'POST',
        body: {...credentials},
      }),
    }),
  }),
  overrideExisting: true,
});

export const {useLoginMutation,useSignupMutation,useLogoutMutation} = authApiSlice || {};