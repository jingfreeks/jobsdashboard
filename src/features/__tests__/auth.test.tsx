import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { configureStore } from '@reduxjs/toolkit';
import { apiSlice } from '@/config/apiSplice';
import authReducer, { 
  setCredentials, 
  setLogout, 
  rehydrateAuth,
  selectCurrentUser,
  selectCurrentUserId,
  selectCurrentToken,
  selectUserRoles,
  selectIsAuthenticated
} from '../auth';
import { authApiSlice } from '../authSlice';
import type { RootState } from '@/config/store';
// Mock console.warn to avoid noise in tests
const mockConsoleWarn = vi.fn();

describe('Auth Feature', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.spyOn(console, 'warn').mockImplementation(mockConsoleWarn);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('Auth Slice', () => {
    let store: ReturnType<typeof createTestStore>;

    const createTestStore = () => {
      return configureStore({
        reducer: {
          auth: authReducer,
          api: apiSlice.reducer,
        },
        middleware: (getDefaultMiddleware) =>
          getDefaultMiddleware().concat(apiSlice.middleware),
      });
    };

    beforeEach(() => {
      store = createTestStore();
    });

    describe('Initial State', () => {
      it('should have correct initial state', () => {
        const state = store.getState();
        
        expect(state.auth).toEqual({
          user: null,
          token: null,
          userId: null,
          roles: [],
          onboardingComplete: false,
        });
      });

      it('should match snapshot for initial state', () => {
        const state = store.getState();
        expect(state.auth).toMatchSnapshot();
      });
    });

    describe('setCredentials Action', () => {
      it('should set credentials correctly with valid payload', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe('testuser@example.com');
        expect(state.auth.token).toBe('valid-token-123');
        expect(state.auth.userId).toBe('user123');
        expect(state.auth.roles).toEqual(['user', 'admin']);
        expect(mockConsoleWarn).not.toHaveBeenCalled();
      });

      it('should handle empty roles array', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: [],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle non-array roles and convert to empty array', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: 'invalid-roles' as unknown as string[],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle missing user and log warning', () => {
        const credentials = {
          user: '',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
        expect(mockConsoleWarn).toHaveBeenCalledWith('Invalid credentials payload:', credentials);
      });

      it('should handle missing accessToken and log warning', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: '',
          userId: 'user123',
          roles: ['user'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
        expect(mockConsoleWarn).toHaveBeenCalledWith('Invalid credentials payload:', credentials);
      });

      it('should handle missing userId and log warning', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: '',
          roles: ['user'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
        expect(mockConsoleWarn).toHaveBeenCalledWith('Invalid credentials payload:', credentials);
      });

      it('should handle null values and log warning', () => {
        const credentials = {
          user: null as unknown as string,
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
        expect(mockConsoleWarn).toHaveBeenCalledWith('Invalid credentials payload:', credentials);
      });

      it('should handle undefined values and log warning', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: undefined as unknown as string,
          userId: 'user123',
          roles: ['user'],
        };

        store.dispatch(setCredentials(credentials));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
        expect(mockConsoleWarn).toHaveBeenCalledWith('Invalid credentials payload:', credentials);
      });

      it('should match snapshot after setting credentials', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };

        store.dispatch(setCredentials(credentials));
        const state = store.getState();
        expect(state.auth).toMatchSnapshot();
      });
    });

    describe('setLogout Action', () => {
      it('should clear all auth state', () => {
        // First set some credentials
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };
        store.dispatch(setCredentials(credentials));

        // Then logout
        store.dispatch(setLogout());

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
      });

      it('should match snapshot after logout', () => {
        // First set some credentials
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };
        store.dispatch(setCredentials(credentials));

        // Then logout
        store.dispatch(setLogout());
        const state = store.getState();
        expect(state.auth).toMatchSnapshot();
      });

      it('should handle logout when already logged out', () => {
        store.dispatch(setLogout());

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
      });
    });

    describe('rehydrateAuth Action', () => {
      it('should rehydrate with partial state', () => {
        const rehydrateData = {
          user: 'testuser@example.com',
          token: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.user).toBe('testuser@example.com');
        expect(state.auth.token).toBe('valid-token-123');
        expect(state.auth.userId).toBe('user123');
        expect(state.auth.roles).toEqual(['user']);
      });

      it('should handle partial rehydration with only user', () => {
        const rehydrateData = {
          user: 'testuser@example.com',
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.user).toBe('testuser@example.com');
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle partial rehydration with only token', () => {
        const rehydrateData = {
          token: 'valid-token-123',
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe('valid-token-123');
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle partial rehydration with only userId', () => {
        const rehydrateData = {
          userId: 'user123',
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe('user123');
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle partial rehydration with only roles', () => {
        const rehydrateData = {
          roles: ['admin', 'moderator'],
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual(['admin', 'moderator']);
      });

      it('should handle non-array roles in rehydration', () => {
        const rehydrateData = {
          roles: 'invalid-roles' as any,
        };

        store.dispatch(rehydrateAuth(rehydrateData));

        const state = store.getState();
        expect(state.auth.roles).toEqual([]);
      });

      it('should handle empty rehydration data', () => {
        store.dispatch(rehydrateAuth({}));

        const state = store.getState();
        expect(state.auth.user).toBe(null);
        expect(state.auth.token).toBe(null);
        expect(state.auth.userId).toBe(null);
        expect(state.auth.roles).toEqual([]);
      });

      it('should match snapshot after rehydration', () => {
        const rehydrateData = {
          user: 'testuser@example.com',
          token: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };

        store.dispatch(rehydrateAuth(rehydrateData));
        const state = store.getState();
        expect(state.auth).toMatchSnapshot();
      });
    });

    describe('State Transitions', () => {
      it('should handle login -> logout -> login cycle', () => {
        // First login
        const credentials1 = {
          user: 'user1@example.com',
          accessToken: 'token1',
          userId: 'user1',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials1));

        let state = store.getState();
        expect(state.auth.user).toBe('user1@example.com');

        // Logout
        store.dispatch(setLogout());
        state = store.getState();
        expect(state.auth.user).toBe(null);

        // Second login
        const credentials2 = {
          user: 'user2@example.com',
          accessToken: 'token2',
          userId: 'user2',
          roles: ['admin'],
        };
        store.dispatch(setCredentials(credentials2));

        state = store.getState();
        expect(state.auth.user).toBe('user2@example.com');
        expect(state.auth.roles).toEqual(['admin']);
      });

      it('should handle multiple rehydrations', () => {
        // First rehydration
        store.dispatch(rehydrateAuth({ user: 'user1@example.com' }));
        let state = store.getState();
        expect(state.auth.user).toBe('user1@example.com');

        // Second rehydration
        store.dispatch(rehydrateAuth({ token: 'token1' }));
        state = store.getState();
        expect(state.auth.user).toBe('user1@example.com');
        expect(state.auth.token).toBe('token1');

        // Third rehydration
        store.dispatch(rehydrateAuth({ userId: 'user1', roles: ['admin'] }));
        state = store.getState();
        expect(state.auth.user).toBe('user1@example.com');
        expect(state.auth.token).toBe('token1');
        expect(state.auth.userId).toBe('user1');
        expect(state.auth.roles).toEqual(['admin']);
      });
    });
  });

  describe('Selectors', () => {
    let store: ReturnType<typeof createTestStore>;

    const createTestStore = () => {
      return configureStore({
        reducer: {
          auth: authReducer,
          api: apiSlice.reducer,
        },
        middleware: (getDefaultMiddleware) =>
          getDefaultMiddleware().concat(apiSlice.middleware),
      });
    };

    // Helper function to get state with proper typing
    //const getState = () => store.getState() as RootState;

    beforeEach(() => {
      store = createTestStore();
    });

    describe('selectCurrentUser', () => {
      it('should return current user', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        expect(selectCurrentUser(state)).toBe('testuser@example.com');
      });

      it('should return null when no user', () => {
        const state = store.getState() as RootState;
        expect(selectCurrentUser(state)).toBe(null);
      });
    });

    describe('selectCurrentUserId', () => {
      it('should return current user ID', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        
        expect(selectCurrentUserId(state)).toBe('user123');
      });

      it('should return null when no user ID', () => {
        const state = store.getState() as RootState;
        expect(selectCurrentUserId(state)).toBe(null);
      });
    });

    describe('selectCurrentToken', () => {
      it('should return current token', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        expect(selectCurrentToken(state)).toBe('valid-token-123');
      });

      it('should return null when no token', () => {
        const state = store.getState() as RootState;
        expect(selectCurrentToken(state)).toBe(null);
      });
    });

    describe('selectUserRoles', () => {
      it('should return user roles', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        expect(selectUserRoles(state)).toEqual(['user', 'admin']);
      });

      it('should return empty array when no roles', () => {
        const state = store.getState() as RootState;
        expect(selectUserRoles(state)).toEqual([]);
      });
    });

    describe('selectIsAuthenticated', () => {
      it('should return true when token exists', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        expect(selectIsAuthenticated(state)).toBe(true);
      });

      it('should return false when no token', () => {
        const state = store.getState() as RootState;
        expect(selectIsAuthenticated(state)).toBe(false);
      });

      it('should return false when token is empty string', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: '',
          userId: 'user123',
          roles: ['user'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        expect(selectIsAuthenticated(state)).toBe(false);
      });
    });

    describe('Selector Snapshots', () => {
      it('should match snapshot for all selectors when authenticated', () => {
        const credentials = {
          user: 'testuser@example.com',
          accessToken: 'valid-token-123',
          userId: 'user123',
          roles: ['user', 'admin'],
        };
        store.dispatch(setCredentials(credentials));

        const state = store.getState() as RootState;
        const selectorResults = {
          currentUser: selectCurrentUser(state),
          currentUserId: selectCurrentUserId(state),
          currentToken: selectCurrentToken(state),
          userRoles: selectUserRoles(state),
          isAuthenticated: selectIsAuthenticated(state),
        };

        expect(selectorResults).toMatchSnapshot();
      });

      it('should match snapshot for all selectors when not authenticated', () => {
        const state = store.getState() as RootState;
        const selectorResults = {
          currentUser: selectCurrentUser(state),
          currentUserId: selectCurrentUserId(state),
          currentToken: selectCurrentToken(state),
          userRoles: selectUserRoles(state),
          isAuthenticated: selectIsAuthenticated(state),
        };

        expect(selectorResults).toMatchSnapshot();
      });
    });
  });

  describe('Auth API Slice', () => {
    it('should have authApiSlice defined', () => {
      expect(authApiSlice).toBeDefined();
    });

    it('should match snapshot for auth API slice', () => {
      expect(authApiSlice).toMatchSnapshot();
    });

    it('should have endpoints property', () => {
      expect(authApiSlice.endpoints).toBeDefined();
    });
  });

  describe('Edge Cases and Error Handling', () => {
    let store: ReturnType<typeof createTestStore>;

    const createTestStore = () => {
      return configureStore({
        reducer: {
          auth: authReducer,
        },
      });
    };

    beforeEach(() => {
      store = createTestStore();
    });

    it('should handle very long user names', () => {
      const longUserName = 'a'.repeat(1000);
      const credentials = {
        user: longUserName,
        accessToken: 'valid-token-123',
        userId: 'user123',
        roles: ['user'],
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.user).toBe(longUserName);
    });

    it('should handle very long tokens', () => {
      const longToken = 'a'.repeat(1000);
      const credentials = {
        user: 'testuser@example.com',
        accessToken: longToken,
        userId: 'user123',
        roles: ['user'],
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.token).toBe(longToken);
    });

    it('should handle very long user IDs', () => {
      const longUserId = 'a'.repeat(1000);
      const credentials = {
        user: 'testuser@example.com',
        accessToken: 'valid-token-123',
        userId: longUserId,
        roles: ['user'],
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.userId).toBe(longUserId);
    });

    it('should handle many roles', () => {
      const manyRoles = Array.from({ length: 100 }, (_, i) => `role${i}`);
      const credentials = {
        user: 'testuser@example.com',
        accessToken: 'valid-token-123',
        userId: 'user123',
        roles: manyRoles,
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.roles).toEqual(manyRoles);
    });

    it('should handle special characters in user names', () => {
      const specialUserName = 'test+user@example.com';
      const credentials = {
        user: specialUserName,
        accessToken: 'valid-token-123',
        userId: 'user123',
        roles: ['user'],
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.user).toBe(specialUserName);
    });

    it('should handle unicode characters', () => {
      const unicodeUserName = 'tëst@exämple.com';
      const credentials = {
        user: unicodeUserName,
        accessToken: 'valid-token-123',
        userId: 'user123',
        roles: ['user'],
      };

      store.dispatch(setCredentials(credentials));

      const state = store.getState();
      expect(state.auth.user).toBe(unicodeUserName);
    });
  });

  describe('Performance and Memory', () => {
    let store: ReturnType<typeof createTestStore>;

    const createTestStore = () => {
      return configureStore({
        reducer: {
          auth: authReducer,
        },
      });
    };

    beforeEach(() => {
      store = createTestStore();
    });

    it('should handle rapid state changes', () => {
      for (let i = 0; i < 100; i++) {
        const credentials = {
          user: `user${i}@example.com`,
          accessToken: `token${i}`,
          userId: `user${i}`,
          roles: [`role${i}`],
        };
        store.dispatch(setCredentials(credentials));
        store.dispatch(setLogout());
      }

      const state = store.getState();
      expect(state.auth.user).toBe(null);
      expect(state.auth.token).toBe(null);
      expect(state.auth.userId).toBe(null);
      expect(state.auth.roles).toEqual([]);
    });

    it('should handle rapid rehydrations', () => {
      for (let i = 0; i < 100; i++) {
        store.dispatch(rehydrateAuth({
          user: `user${i}@example.com`,
          token: `token${i}`,
          userId: `user${i}`,
          roles: [`role${i}`],
        }));
      }

      const state = store.getState();
      expect(state.auth.user).toBe('user99@example.com');
      expect(state.auth.token).toBe('token99');
      expect(state.auth.userId).toBe('user99');
      expect(state.auth.roles).toEqual(['role99']);
    });
  });
}); 