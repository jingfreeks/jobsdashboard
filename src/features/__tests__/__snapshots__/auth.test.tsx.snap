// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auth Feature > Auth API Slice > should match snapshot for auth API slice 1`] = `
{
  "endpoints": {
    "login": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "login",
      "select": [Function],
      "useMutation": [Function],
    },
    "logout": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "logout",
      "select": [Function],
      "useMutation": [Function],
    },
    "signup": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "signup",
      "select": [Function],
      "useMutation": [Function],
    },
  },
  "enhanceEndpoints": [Function],
  "injectEndpoints": [Function],
  "internalActions": {
    "cacheEntriesUpserted": [Function],
    "internal_getRTKQSubscriptions": [Function],
    "middlewareRegistered": [Function],
    "onFocus": [Function],
    "onFocusLost": [Function],
    "onOffline": [Function],
    "onOnline": [Function],
    "queryResultPatched": [Function],
    "removeMutationResult": [Function],
    "removeQueryResult": [Function],
    "resetApiState": [Function],
    "subscriptionsUpdated": [Function],
    "unsubscribeQueryResult": [Function],
    "updateProvidedBy": [Function],
    "updateSubscriptionOptions": [Function],
  },
  "middleware": [Function],
  "reducer": [Function],
  "reducerPath": "api",
  "useLoginMutation": [Function],
  "useLogoutMutation": [Function],
  "usePrefetch": [Function],
  "useSignupMutation": [Function],
  "util": {
    "getRunningMutationThunk": [Function],
    "getRunningMutationsThunk": [Function],
    "getRunningQueriesThunk": [Function],
    "getRunningQueryThunk": [Function],
    "invalidateTags": [Function],
    "patchQueryData": [Function],
    "prefetch": [Function],
    "resetApiState": [Function],
    "selectCachedArgsForQuery": [Function],
    "selectInvalidatedBy": [Function],
    "updateQueryData": [Function],
    "upsertQueryData": [Function],
    "upsertQueryEntries": [Function],
  },
}
`;

exports[`Auth Feature > Auth Slice > Initial State > should match snapshot for initial state 1`] = `
{
  "roles": [],
  "token": null,
  "user": null,
  "userId": null,
}
`;

exports[`Auth Feature > Auth Slice > rehydrateAuth Action > should match snapshot after rehydration 1`] = `
{
  "roles": [
    "user",
    "admin",
  ],
  "token": "valid-token-123",
  "user": "testuser@example.com",
  "userId": "user123",
}
`;

exports[`Auth Feature > Auth Slice > setCredentials Action > should match snapshot after setting credentials 1`] = `
{
  "roles": [
    "user",
    "admin",
  ],
  "token": "valid-token-123",
  "user": "testuser@example.com",
  "userId": "user123",
}
`;

exports[`Auth Feature > Auth Slice > setLogout Action > should match snapshot after logout 1`] = `
{
  "roles": [],
  "token": null,
  "user": null,
  "userId": null,
}
`;

exports[`Auth Feature > Selectors > Selector Snapshots > should match snapshot for all selectors when authenticated 1`] = `
{
  "currentToken": "valid-token-123",
  "currentUser": "testuser@example.com",
  "currentUserId": "user123",
  "isAuthenticated": true,
  "userRoles": [
    "user",
    "admin",
  ],
}
`;

exports[`Auth Feature > Selectors > Selector Snapshots > should match snapshot for all selectors when not authenticated 1`] = `
{
  "currentToken": null,
  "currentUser": null,
  "currentUserId": null,
  "isAuthenticated": false,
  "userRoles": [],
}
`;
